rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,367 @@
+compute resource processing the job.
+
+##### S5 -> S6
+
+Once the upload has been canceled by the compute resource the state will be canceled.
+
+##### S2 -> S3
+
+The upload has been completed. The state of the upload is uploaded.
+
+##### S2 -> S1
+
+If the upload fails, the job returns to the queue to be retried at a later time. The state changes to waiting.
+
+#### S2 -> S4
+
+The upload job has failed to download after several retries. The state of the upload is failed.
+
+### Test plan
+
+1. Verify that a large file can be uploaded using a shared link from Dropbox.
+1. Verify that a small file can be uploaded using a shared link from Dropbox.
+1. Verify the upload queue is not cleared until the file is store in S3.
+1. Verify a failed upload is marked as failed.
+1. Verify an upload is retried if it fails.
+1. Verify only the owner can query the status of an upload.
+1. Verify only the owner can upload toa dataset to their collection.
+1. Verify partial uploads are erased from S3 if they fail or are canceled.
+1. Verify a user can start another upload if the original upload failed, or was canceled.
+1. Verify a shared folder from DropBox cannot be uploaded.
+1. Verify uploaded files are removed from the upload bucket, once they moved to the primary bucket.
+1. Verify an uploaded file can be overwritten.
+1. Verify only [accepted file types](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations)
+   are allowed.
+
+### Monitoring and error reporting
+
+- Monitor the length of the upload queue. If it gets too long, then increase the number of compute instances used.
+- Log failure on the compute resource.
+
+## [Optional] Alternatives
+
+### Download using SMB
+
+Dropbox and Google drive support syncing files to a server using their desktop application. This used SMB protocol
+to sync your local storage with what is in the cloud. This will allow a user to share a file with us by adding us
+to their shared folder. This would cause the file to sync to our system. This approach would solve the download problem,
+and at the same time introduce more uploading and cleanup problems. The download to our server would be asynchronous,
+which would mean the server needs to be up 100% of the time. Instruction must be provided to users on how to share
+their files with our account. The file names would be controlled by the user until they are synced into our server.
+From that point, a mapping of files to their file name uploaded to S3 would be needed. Given a cluster of servers, I'm not
+certain if the SMB protocol would duplicate the files across all of the servers or if the servers can be instructed which
+files to clone.
+
+## References
+
+Any relevant documents or prior art should be clearly listed here. Any citation of figures or papers should also be clearly listed here. Of course, they can also be linked throughout the document when directly referenced.
```

Comment:
Can delete this line; it's part of the template.

Reply:
removed